File: frontend/src/services/timetable/generate.ts
import api from "../../lib/api";

export interface GenerationGap {
  subject_name: string;
  teacher_name?: string;
  reason: string;
}

export interface ClassGenerationResult {
  class_id: string;
  class_name: string;
  timetable_id?: string;
  placed: number;
  required: number;
  gaps: GenerationGap[];
  skipped: boolean;
  skip_reason?: string;
}

export interface GenerationResult {
  classes: ClassGenerationResult[];
}

export interface GenerateTimetablesRequest {
  grade_section_id: string;
  academic_year_id: string;
}

export const generateApi = {
  generate: (data: GenerateTimetablesRequest) =>
    api.post<GenerationResult>("/timetables/generate", data).then((r) => r.data),
};

File: backend/internal/models/timetable/generate.go
package timetable

import "github.com/google/uuid"

type GenerateTimetablesRequest struct {
	GradeSectionID uuid.UUID `json:"grade_section_id" binding:"required"`
	AcademicYearID uuid.UUID `json:"academic_year_id" binding:"required"`
}

// GenerationGap is one period-instance the generator couldn't place, and why.
type GenerationGap struct {
	SubjectName string `json:"subject_name"`
	TeacherName string `json:"teacher_name,omitempty"`
	Reason      string `json:"reason"`
}

type ClassGenerationResult struct {
	ClassID     uuid.UUID       `json:"class_id"`
	ClassName   string          `json:"class_name"`
	TimetableID *uuid.UUID      `json:"timetable_id,omitempty"`
	Placed      int             `json:"placed"`
	Required    int             `json:"required"`
	Gaps        []GenerationGap `json:"gaps"`
	// Skipped is true when the class already has a non-draft timetable this
	// year — it's left untouched rather than regenerated.
	Skipped    bool   `json:"skipped"`
	SkipReason string `json:"skip_reason,omitempty"`
}

type GenerationResult struct {
	Classes []ClassGenerationResult `json:"classes"`
}
